File: BaekJoon/분할 정복/2630_색종이 만들기.cpp
#include <iostream>
using namespace std;

int n;
int square[128][128];
int white, blue;

void solve(int row, int col, int size){
    int check = square[row][col];
    
    for (int i = row; i < row + size; i++){
        for (int j = col; j < col + size; j++){
            if (square[i][j] != check){
                solve(row, col, size/2);
                solve(row, col + size/2, size/2);
                solve(row + size/2, col, size/2);
                solve(row + size/2, col + size/2, size/2);
                return;
            }
        }
    }
    
    if (check == 0){
        white++;
    }
    else{
        blue++;
    }
}

int main(){
    cin >> n;
    
    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            cin >> square[i][j];
        }
    }
    
    solve(0, 0, n);
    cout << white << endl;
    cout << blue << endl;
}

File: BaekJoon/분할 정복/2630_색종이 만들기.swift
// 나눌 수 없을 때까지 나눈다.
let n = Int(readLine()!)!
var papers = [[Int]]()
var blue = 0
var white = 0

for _ in 0..<n {
    papers.append(readLine()!.split(separator: " ").map { Int(String($0))! })
}

divide(0, 0, n)

print(white)
print(blue)

func divide(_ row: Int, _ col: Int, _ size: Int) {
    let color = papers[row][col] // 시작 지점의 색상
    
    for i in row..<row + size {
        for j in col..<col + size {
            if papers[i][j] != color { // 기준 색상과 다르면, 분할
                divide(row, col, size / 2)
                divide(row, col + size / 2, size / 2)
                divide(row + size / 2, col, size / 2)
                divide(row + size / 2, col + size / 2, size / 2)
                return
            }
        }
    }
    
    if color == 1 {
        blue += 1
    } else {
        white += 1
    }
}